src/Type/Php/CurlInitReturnTypeExtension.php

```diff
@@ -26,13 +45,58 @@ public function getTypeFromFunctionCall(
 		Scope $scope,
 	): Type
 	{
-		$argsCount = count($functionCall->getArgs());
+		$args = $functionCall->getArgs();
+		$argsCount = count($args);
 		$returnType = ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
+		$notFalseReturnType = TypeCombinator::remove($returnType, new ConstantBooleanType(false));
 		if ($argsCount === 0) {
-			return TypeCombinator::remove($returnType, new ConstantBooleanType(false));
+			return $notFalseReturnType;
+		}
+
+		$urlArgType = $scope->getType($args[0]->value);
+		if ($urlArgType->isConstantScalarValue()->yes() && (new UnionType([new NullType(), new StringType()]))->isSuperTypeOf($urlArgType)->yes()) {
+			$urlArgReturnTypes = array_map(
+				fn ($value) => $this->getUrlArgValueReturnType($value, $returnType, $notFalseReturnType),
+				$urlArgType->getConstantScalarValues(),
+			);
+			return TypeCombinator::union(...$urlArgReturnTypes);
 		}
 
 		return $returnType;
 	}
 
+	private function getUrlArgValueReturnType(mixed $urlArgValue, Type $returnType, Type $notFalseReturnType): Type
+	{
+		if ($urlArgValue === null) {
+			return $notFalseReturnType;
+		}
+		if (!is_string($urlArgValue)) {
+			throw new ShouldNotHappenException();
+		}
+		if (str_contains($urlArgValue, "\0")) {
+			if (!$this->phpVersion->throwsValueErrorForInternalFunctions()) {
+				// https://github.com/php/php-src/blob/php-7.4.33/ext/curl/interface.c#L112-L115
+				return new ConstantBooleanType(false);
+			}
+			// https://github.com/php/php-src/blob/php-8.0.0/ext/curl/interface.c#L104-L107
+			return new NeverType();
+		}
+		if ($this->phpVersion->getVersionId() < 80000) {
```

Comment:
Should be a new PHPVersion method

Reply:
Done (`isCurlOptUrlCheckingFileSchemeWithOpenBasedir()`)